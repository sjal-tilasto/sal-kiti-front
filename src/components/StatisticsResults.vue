<template>
  <div>
    <b-row>
      <b-col>
        <h3 v-if="header" class="bg-sal-orange">
          {{ $tc("result.result", 2) }}: {{ header }}
        </h3>
        <h3 v-else class="bg-sal-orange">{{ $tc("result.result", 2) }}</h3>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-alert v-if="errors.main" variant="danger" show>
          <ul>
            <li v-for="e in errors.main" v-bind:key="e">{{ e }}</li>
          </ul>
        </b-alert>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-pagination
          v-model="currentPage"
          :total-rows="results.count"
          :per-page="results.limit"
          align="right"
          v-if="results.count > results.limit"
        >
        </b-pagination>
        <p class="text-right" v-if="results.page">
          {{ $t("search.count") }}: {{ results.count }}
        </p>
        <b-table
          v-if="results.count > 0"
          small
          :fields="resultFields"
          :items="results.results"
        >
          <template v-slot:cell(number)="data">
            <div
              v-if="
                !results.results[data.index - 1] ||
                  data.item.result !== results.results[data.index - 1].result
              "
            >
              <div
                v-if="
                  highlight &&
                    data.index +
                      1 +
                      results.limit * currentPage -
                      results.limit <=
                      highlight
                "
              >
                <span class="text-success">{{
                  data.index + 1 + results.limit * currentPage - results.limit
                }}</span>
              </div>
              <div v-else>
                {{
                  data.index + 1 + results.limit * currentPage - results.limit
                }}
              </div>
            </div>
          </template>
          <template v-slot:cell(athlete)="data">
            <div v-if="data.item.athlete">
              <router-link
                tag="a"
                :to="{
                  name: 'athlete',
                  params: { athlete_id: data.item.athlete.id }
                }"
              >
                {{ data.item.athlete.first_name }}
                {{ data.item.athlete.last_name }}
              </router-link>
            </div>
            <div v-else>
              {{ $t("team") }}: {{ data.item.last_name }}
              <ul>
                <li v-for="member in data.item.team_members" :key="member.id">
                  <router-link
                    tag="a"
                    :to="{
                      name: 'athlete',
                      params: { athlete_id: member.id }
                    }"
                  >
                    {{ member.first_name }} {{ member.last_name }}
                  </router-link>
                </li>
              </ul>
            </div>
          </template>
          <template v-slot:cell(result)="data">
            {{ data.item.result | roundValue(data.item.decimals) }}
          </template>
          <template v-slot:cell(competition)="data">
            <router-link
              tag="a"
              :to="{
                name: 'competition',
                params: { competition_id: data.item.competition.id }
              }"
            >
              {{ data.item.competition.name }}
            </router-link>
          </template>
          <template v-slot:cell(date)="data">
            <div
              v-if="
                data.item.competition.date_start ===
                  data.item.competition.date_end
              "
            >
              {{ data.item.competition.date_start }}
            </div>
            <div v-else>
              {{ data.item.competition.date_start }} -
              {{ data.item.competition.date_end }}
            </div>
          </template>
        </b-table>
        <div v-show="loadingResults">
          <b-spinner label="Loading..."></b-spinner>
        </div>
      </b-col>
    </b-row>
  </div>
</template>

<script>
import { HTTP } from "../api/BaseApi.js";
import roundValue from "../utils/RoundValueFilter";
import errorParser from "../utils/ErrorParser";

export default {
  name: "Statistics",
  filters: {
    roundValue
  },
  props: {
    header: {
      type: String,
      default: null
    },
    highlight: {
      type: Number,
      default: null
    },
    searchParameters: String
  },
  data() {
    return {
      currentPage: 1,
      errors: {},
      loadingResults: false,
      results: []
    };
  },
  computed: {
    /**
     * Sets fields list for the statistics list
     *
     * @returns {array} fields list
     */
    resultFields: function() {
      let groupResult = this.searchParameters.includes("group_results");
      let fields = [
        { key: "number", label: "#" },
        { key: "athlete", label: this.$tc("athlete.athlete", 1) }
      ];
      if (!groupResult) {
        fields.push({
          key: "organization",
          label: this.$t("athlete.club"),
          thClass: "d-none d-md-table-cell",
          tdClass: "d-none d-md-table-cell"
        });
        fields.push({ key: "category", label: this.$t("result.category") });
      }
      if (groupResult) {
        fields.push({
          key: "athlete.organization_info.abbreviation",
          label: this.$tc("athlete.club", 1)
        });
      }
      fields.push({ key: "result", label: this.$tc("result.result", 1) });
      if (!groupResult) {
        fields.push({
          key: "competition.type_info.abbreviation",
          label: this.$t("competition.type")
        });
        fields.push({
          key: "competition.level_info.abbreviation",
          label: this.$t("competition.level"),
          thClass: "d-none d-md-table-cell",
          tdClass: "d-none d-md-table-cell"
        });
        fields.push({
          key: "competition",
          label: this.$tc("competition.competition", 1),
          thClass: "d-none d-lg-table-cell",
          tdClass: "d-none d-lg-table-cell"
        });
        fields.push({
          key: "date",
          label: this.$t("date"),
          thClass: "d-none d-lg-table-cell",
          tdClass: "d-none d-lg-table-cell"
        });
      }
      return fields;
    }
  },
  watch: {
    /**
     * Update results on page change
     */
    currentPage: {
      handler: function() {
        this.getStatistics();
      }
    },
    /**
     * Trigger search on parameters change
     */
    searchParameters: {
      handler: function() {
        this.currentPage = 1;
        this.getStatistics();
      }
    }
  },
  mounted() {
    this.getStatistics();
  },
  methods: {
    /**
     * Fetch statistics from API
     *
     * @returns {Promise<void>}
     */
    async getStatistics() {
      if (this.currentPage) {
        if (
          !this.results.count ||
          this.results.count <= (this.currentPage - 1) * this.limit
        ) {
          this.currentPage = 1;
        }
      }
      this.loadingResults = true;
      this.results = [];
      this.$set(this.errors, "main", null);
      HTTP.get(
        "resultlist/" + this.searchParameters + "&page=" + this.currentPage
      )
        .then(response => {
          this.results = response.data || [];
        })
        .catch(error => {
          this.$set(this.errors, "main", errorParser.generic.bind(this)(error));
        })
        .finally(() => (this.loadingResults = false));
    }
  }
};
</script>

<style scoped></style>
