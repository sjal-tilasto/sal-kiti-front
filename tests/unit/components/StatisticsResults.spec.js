import mockAxios from "axios";
import {
  createLocalVue,
  mount,
  RouterLinkStub,
  shallowMount
} from "@vue/test-utils";
import flushPromises from "flush-promises";
import StatisticsResults from "@/components/StatisticsResults.vue";
import BootstrapVue from "bootstrap-vue";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(BootstrapVue);
localVue.use(Vuex);

describe("StatisticsResults.vue", () => {
  it("renders result list", async () => {
    const wrapper = mount(StatisticsResults, {
      localVue,
      stubs: { "router-link": RouterLinkStub },
      propsData: {
        searchParameters:
          "?fields!=partial&category=1&level=2&type=1,2&organization=1&start=20190608&end=20190713&limit=10"
      }
    });
    await flushPromises();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("render error message on error", async () => {
    mockAxios.get.mockImplementationOnce(() =>
      Promise.reject({
        response: {
          status: 404,
          statusText: "Not Found"
        }
      })
    );
    const wrapper = shallowMount(StatisticsResults, {
      localVue,
      propsData: {
        searchParameters: "?fields!=partial&limit=10"
      }
    });
    await flushPromises();
    expect(wrapper.text()).toContain("import.error.notfound");
  });
});
