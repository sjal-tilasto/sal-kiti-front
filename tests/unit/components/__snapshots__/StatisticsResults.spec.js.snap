// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticsResults.vue renders result list 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <h3
        class="bg-sal-orange"
      >
        result.result
      </h3>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col"
    >
      <!---->
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col"
    >
      <!---->
       
      <p
        class="text-right"
      >
        
        search.count: 2
      
      </p>
       
      <table
        aria-busy="false"
        aria-colcount="9"
        class="table b-table table-sm"
        id="__BVID__7"
        role="table"
      >
        <!---->
        <!---->
        <thead
          class=""
          role="rowgroup"
        >
          <!---->
          <tr
            class=""
            role="row"
          >
            <th
              aria-colindex="1"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                #
              </div>
            </th>
            <th
              aria-colindex="2"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                athlete.athlete
              </div>
            </th>
            <th
              aria-colindex="3"
              class="d-none d-md-table-cell"
              role="columnheader"
              scope="col"
            >
              <div>
                athlete.club
              </div>
            </th>
            <th
              aria-colindex="4"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                result.category
              </div>
            </th>
            <th
              aria-colindex="5"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                result.result
              </div>
            </th>
            <th
              aria-colindex="6"
              class=""
              role="columnheader"
              scope="col"
            >
              <div>
                competition.type
              </div>
            </th>
            <th
              aria-colindex="7"
              class="d-none d-md-table-cell"
              role="columnheader"
              scope="col"
            >
              <div>
                competition.level
              </div>
            </th>
            <th
              aria-colindex="8"
              class="d-none d-lg-table-cell"
              role="columnheader"
              scope="col"
            >
              <div>
                competition.competition
              </div>
            </th>
            <th
              aria-colindex="9"
              class="d-none d-lg-table-cell"
              role="columnheader"
              scope="col"
            >
              <div>
                date
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <!---->
          <tr
            class=""
            role="row"
          >
            <td
              aria-colindex="1"
              class=""
              role="cell"
            >
              <div>
                <div>
                  
              1
            
                </div>
              </div>
            </td>
            <td
              aria-colindex="2"
              class=""
              role="cell"
            >
              <div>
                <a>
                  
              Sara
              Huffy
            
                </a>
              </div>
            </td>
            <td
              aria-colindex="3"
              class="d-none d-md-table-cell"
              role="cell"
            >
              Club
            </td>
            <td
              aria-colindex="4"
              class=""
              role="cell"
            >
              N
            </td>
            <td
              aria-colindex="5"
              class=""
              role="cell"
            >
              
          200,6
        
            </td>
            <td
              aria-colindex="6"
              class=""
              role="cell"
            >
              50MC
            </td>
            <td
              aria-colindex="7"
              class="d-none d-md-table-cell"
              role="cell"
            >
              NC
            </td>
            <td
              aria-colindex="8"
              class="d-none d-lg-table-cell"
              role="cell"
            >
              <a>
                
            Demo Comp
          
              </a>
            </td>
            <td
              aria-colindex="9"
              class="d-none d-lg-table-cell"
              role="cell"
            >
              <div>
                
            2019-01-23
          
              </div>
            </td>
          </tr>
          <tr
            class=""
            role="row"
          >
            <td
              aria-colindex="1"
              class=""
              role="cell"
            >
              <div>
                <div>
                  
              2
            
                </div>
              </div>
            </td>
            <td
              aria-colindex="2"
              class=""
              role="cell"
            >
              <div>
                <a>
                  
              Phyllis
              Miller
            
                </a>
              </div>
            </td>
            <td
              aria-colindex="3"
              class="d-none d-md-table-cell"
              role="cell"
            >
              ABC
            </td>
            <td
              aria-colindex="4"
              class=""
              role="cell"
            >
              N
            </td>
            <td
              aria-colindex="5"
              class=""
              role="cell"
            >
              
          206,4
        
            </td>
            <td
              aria-colindex="6"
              class=""
              role="cell"
            >
              50MC
            </td>
            <td
              aria-colindex="7"
              class="d-none d-md-table-cell"
              role="cell"
            >
              NC
            </td>
            <td
              aria-colindex="8"
              class="d-none d-lg-table-cell"
              role="cell"
            >
              <a>
                
            Demo Comp
          
              </a>
            </td>
            <td
              aria-colindex="9"
              class="d-none d-lg-table-cell"
              role="cell"
            >
              <div>
                
            2019-01-23
          
              </div>
            </td>
          </tr>
          <tr
            class=""
            role="row"
          >
            <td
              aria-colindex="1"
              class=""
              role="cell"
            >
              <div>
                <div>
                  
              3
            
                </div>
              </div>
            </td>
            <td
              aria-colindex="2"
              class=""
              role="cell"
            >
              <div>
                <a>
                  
              Jessica
              Carlson
            
                </a>
              </div>
            </td>
            <td
              aria-colindex="3"
              class="d-none d-md-table-cell"
              role="cell"
            >
              ABC
            </td>
            <td
              aria-colindex="4"
              class=""
              role="cell"
            >
              M
            </td>
            <td
              aria-colindex="5"
              class=""
              role="cell"
            >
              
          222
        
            </td>
            <td
              aria-colindex="6"
              class=""
              role="cell"
            >
              50MC
            </td>
            <td
              aria-colindex="7"
              class="d-none d-md-table-cell"
              role="cell"
            >
              NC
            </td>
            <td
              aria-colindex="8"
              class="d-none d-lg-table-cell"
              role="cell"
            >
              <a>
                
            Demo Comp
          
              </a>
            </td>
            <td
              aria-colindex="9"
              class="d-none d-lg-table-cell"
              role="cell"
            >
              <div>
                
            2019-01-23
          
              </div>
            </td>
          </tr>
          <!---->
          <!---->
        </tbody>
        <!---->
      </table>
       
      <div
        style="display: none;"
      >
        <span
          class="spinner-border"
          role="status"
        >
          <span
            class="sr-only"
          >
            Loading...
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
